BepInEx.Preloader.Core/Patching/AssemblyPatcher.cs

```diff
@@ -412,10 +167,69 @@ public static bool TryLoadAssembly(string path, out AssemblyDefinition assembly)
         }
     }
 
+    /// <inheritdoc />
+    protected override PatcherPlugin LoadPlugin(PluginInfo pluginInfo, Assembly pluginAssembly)
+    {
+        var instance = base.LoadPlugin(pluginInfo, pluginAssembly);
+        instance.Context = PatcherContext;
+
+        PatcherContext.PatcherPlugins.Add(instance);
+        var type = pluginAssembly.GetType(pluginInfo.TypeName);
+        var methods = type.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+        var patchDefinitions = new List<PatchDefinition>();
+        foreach (var method in methods)
+        {
+            var targetAssemblies = MetadataHelper.GetAttributes<TargetAssemblyAttribute>(method);
+            var targetTypes = MetadataHelper.GetAttributes<TargetTypeAttribute>(method);
+
+            if (targetAssemblies.Length == 0 && targetTypes.Length == 0)
+                continue;
+
+            var parameters = method.GetParameters();
+
+            if (parameters.Length < 1 || parameters.Length > 2
+                                         // Next few lines ensure that the first parameter is AssemblyDefinition and does not have any
+                                         // target type attributes, and vice versa
+                                      || !(
+                                              parameters[0].ParameterType == typeof(AssemblyDefinition)
+                                           || parameters[0].ParameterType ==
+                                              typeof(AssemblyDefinition).MakeByRefType()
+                                           && targetTypes.Length == 0
+                                           || parameters[0].ParameterType == typeof(TypeDefinition)
+                                           && targetAssemblies.Length == 0
+                                          )
+                                      || parameters.Length == 2 &&
+                                         parameters[1].ParameterType != typeof(string)
+                                      || method.ReturnType != typeof(void) &&
+                                         method.ReturnType != typeof(bool)
+               )
```

Comment:
This is awful ;)

Reply:
As discussed, a planned refactor that changes the patcher api will remove this code so it can temporarily stay for now